_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,440 @@
+---
+title: "Clean Unit Tests with Mockito"
+categories: [java]
+date: 2021-03-20 00:00:00 +0100 
+modified: 2021-03-20 00:00:00 +0100 
+author: lleuenberger 
+excerpt: "Clean Unit Tests with Mockito"
+image:
+  auto: 0098-profile
+---
+
+In this article we will learn how to mock objects with Mockito. Therefore, we first need to understand what test doubles
+are and then how we can use them so that we create meaningful and tailored unit tests. We will also have a look at the
+most important Dos and Dont's in such way that we are able to write our own unit tests with Mockito.
+
+{% include github-project.html url="https://github.com/silenum/mockito-examples" %}
+
+## Introduction to Mocks
+
+The basic concept of mocking is replacing real objects with doubles, that we are able to advise how to behave. So to
+say, mocks are stuntmen for our test cases. Each stuntman has his speciality at what he is good at, and so do mocks. In
+software engineering we differentiate in **Test Doubles**. Don't get confused with *Mocks* and *Test Doubles*, we will
+cover the difference in detail later in this article.
+
+Let's imagine we have a service, that processes orders from a database. It's very unhandy to set up a whole database
+just to test that service. In that case we create a *mock* that pretends to be the database, but in the eyes from the
+service it looks like a real database. Furthermore, we can exactly advise the mock how it shall behave. That's exactly
+what stuntmen do. They do the hard and cumbersome work, while we can enjoy doing the nice work! Having this tool, we
+are able to test the service but don't actually need a database.
+
+Here [Mockito](https://www.mockito.org) comes into play. Mockito is a very popular library that supports such testing
+scenarios. Spring also uses Mockito and has its own adapters that provide convenience for developers. Don't worry -
+we don't need to have Spring to use Mockito! Mockito also comes with a standalone library that can be referenced within
+our build tool such as Maven and Gradle.
+
+Consider reading the section [Why Mock?](https://reflectoring.io/spring-boot-mock/#why-mock) for additional information
+about Mocking.
+
+## Different Types of Test Doubles
+
+In the world of code, there are lots of words for test doubles and definitions for their duty. I recommend to
+define a common language within the team.
+
+Here is a little summary of the different types for test doubles and how we use them:
+
+| Type  | Description                          |
+| ----- | ------------------------------------------------------------ |
+| Dummy | A dummy is an object that is often used as a parameter for the method under test but without an actual need of it during the test itself. |
+| Fake  | A fake is an object that has an implementation, but which is restricted and simpler as the real implementation. Due to this restriction, the object is not meant to be used in production. |
```

Comment:
I removed "Dummy" and "Fake" from this list because we're not using those words anywhere in the article. I left "Stub", even though it's not used, because it's very similar to "Mock" and often gets confused with it.